gtsfm/common/gtsfm_data.py

```diff
@@ -0,0 +1,199 @@
+
+        Returns:
+            List of indices with a valid camera.
+        """
+        return list(self._cameras.keys())
+
+    def get_camera(self, index: int) -> Optional[PinholeCameraCal3Bundler]:
+        """Getter for camera.
+
+        Args:
+            index: the image index to fetch the camera for.
+
+        Returns:
+            The camera if it is a valid one, None otherwise.
+        """
+        return self._cameras.get(index)
+
+    def get_track(self, index: int) -> SfmTrack:
+        """Getter for the track.
+
+        Args:
+            index: track index to fetch.
+
+        Returns:
+            Requested track.
+        """
+        return self._tracks[index]
+
+    def add_track(self, track: SfmTrack) -> bool:
+        """Add a track, after checking if all the cameras in the track are already added.
+
+        Args:
+            track: track to add.
+
+        Returns:
+            Flag indicating the success of adding operation.
+        """
+        # check if all cameras are already added
+        for j in range(track.number_measurements()):
+            i, _ = track.measurement(j)
+
+            if i not in self._cameras:
+                return False
+
+        self._tracks.append(track)
+        return True
+
+    def add_camera(self, index: int, camera: PinholeCameraCal3Bundler,) -> None:
+        """Adds a camera.
+
+        Args:
+            index: the index associated with this camera.
+            camera: camera object to it.
+
+        Raises:
+            ValueError: if the camera to be added is not a valid camera object.
+        """
+        if camera is None:
+            raise ValueError("Camera cannot be None, should be a valid camera")
+        self._cameras[index] = camera
+
+    def select_largest_connected_component(self) -> "GtsfmData":
+        """Selects the subset of data belonging to the largest connected component of the graph where the edges are
+        between cameras which feature in the same track.
+
+        Returns:
+            New GtSfmData object with the subset of tracks and cameras.
+        """
+        camera_edges = []
+        for sfm_track in self._tracks:
+            cameras_in_use = []
+            for m_idx in range(sfm_track.number_measurements()):
+                i, _ = sfm_track.measurement(m_idx)
+                cameras_in_use.append(i)
+
+            # Recreate track connectivity from track information
+            # For example: a track has cameras [0, 2, 5]. In that case we will add pairs (0, 2), (0, 5), (2, 5)
+            camera_edges += list(itertools.combinations(cameras_in_use, 2))
+
+        if len(camera_edges) == 0:
+            return GtsfmData(self._number_images)
+
+        cameras_in_largest_cc = graph_utils.get_nodes_in_largest_connected_component(camera_edges)
+        return GtsfmData.pick_cameras(self, cameras_in_largest_cc)
+
+    @classmethod
+    def pick_cameras(cls, gtsfm_data: "GtsfmData", camera_indices: List[int]) -> "GtsfmData":
```

Comment:
looks great @ayushbaid . much better. only thing i might change is to name this
`def from_picked_cameras()`

I think class methods often start with `from_blah` to make their factory functionality clearer?

Reply:
updated